### PR TITLE
Add support to obtain the current EventLoop

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

In certain cases it would be useful to be able to access the current EventLoop (which is tied to a thread).

Modifications:

- Add MultiThreadEventLoopGroup.currentEventLoop
- Add tests

Result:

Fixes [#59]

### CODE REVIEW

**Comment by normanmaurer** [comment]
Also @tanner0101

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -660,6 +662,9 @@ final public class MultiThreadedEventLoopGroup: EventLoopGroup {
                 do {
                     /* we try! this as this must work (just setting up kqueue/epoll) or else there's not much we can do here */
                     let l = try! SelectableEventLoop(thread: t)
+                    
+                    threadSpecificEventLoop.currentValue = l
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: Probably we can remove the whitespace above this line.

Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: So the core problem here right now is that this leaks each event loop. The TSV does not (and indeed cannot) free the data stored in it when the thread exits, so this will keep the loop in memory forever.

When the loop is cleaned up, we need to nil out this TSV.

Comment (normanmaurer) [reply Sources/NIO/EventLoop.swift]: Ah I thought it will... that’s easy to fix. Let me do it later

**Comment by normanmaurer** [comment]
@Lukasa all fixed

File: Tests/NIOTests/EventLoopTest.swift [hunk 2]
```diff
@@ -228,4 +228,24 @@ public class EventLoopTest : XCTestCase {
...
+        
+        let loop2 = group.next()
+        let currentLoop2 = try loop2.submit {
+            return MultiThreadedEventLoopGroup.currentEventLoop
+        }.wait()
+        XCTAssertTrue(loop2 === currentLoop2)
+        XCTAssertFalse(loop1 === loop2)
+        
+        XCTAssertNil(MultiThreadedEventLoopGroup.currentEventLoop)
+        XCTAssertNoThrow(try group.syncShutdownGracefully())
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift]: Can you add another test that grabs a weak ref to the event loop and confirm that it doesn't leak?

Comment (normanmaurer) [reply Tests/NIOTests/EventLoopTest.swift]: Good idea... will do.

**Comment by normanmaurer** [comment]
@Lukasa adjusted the test... PTAL again

File: Tests/NIOTests/EventLoopTest.swift [hunk 3]
```diff
@@ -228,4 +228,44 @@ public class EventLoopTest : XCTestCase {
...
+            XCTAssertNotNil(holder.loop)
+            XCTAssertNil(MultiThreadedEventLoopGroup.currentEventLoop)
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+            return holder
+        }
+        
+        let holder = try testCurrentEventLoop0()
+        
+        // Check we not leak the EventLoop.
+        while holder.loop != nil {
```
Comment (normanmaurer) [inline Tests/NIOTests/EventLoopTest.swift]: @Lukasa I had to use a while loop here as it sometimes not directly "clear" the weak reference when `testCurrentEventLoop0` returns.

File: Tests/NIOTests/EventLoopTest.swift [hunk 4]
```diff
@@ -228,4 +228,44 @@ public class EventLoopTest : XCTestCase {
...
+            return holder
+        }
+        
+        let holder = try assertCurrentEventLoop0()
+        
+        // Check we not leak the EventLoop.
+        while holder.loop != nil {
+            // Give the ARC some time to kick in and remove the weak reference.
+            sleep(1)
+        }
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift]: Is this loop actually necessary? ARC should be fully deterministic here if we're using `syncShutdownGracefully`.

Comment (normanmaurer) [reply Tests/NIOTests/EventLoopTest.swift]: Without the whole it sometimes fails 🤷‍♂️

File: Tests/NIOTests/EventLoopTest.swift [hunk 5]
```diff
@@ -228,4 +228,44 @@ public class EventLoopTest : XCTestCase {
...
+            let holder = EventLoopHolder(loop2)
+            XCTAssertNotNil(holder.loop)
+            XCTAssertNil(MultiThreadedEventLoopGroup.currentEventLoop)
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+            return holder
+        }
+        
+        let holder = try assertCurrentEventLoop0()
+        
+        // Check we not leak the EventLoop.
```
Comment (weissi) [inline Tests/NIOTests/EventLoopTest.swift]: grammar nit: "Check we don't leak the EventLoop"

File: Tests/NIOTests/EventLoopTest.swift (hunk 4 above)
Comment (weissi) [reply Tests/NIOTests/EventLoopTest.swift]: so sleeping should not fix ARC issues as it's deterministic as @Lukasa says. The problem here is the race between the `MultithreadedEventLoop`'s thread and the main thread running the test. So I think we should fix the comment and then merge.

Comment (Lukasa) [reply Tests/NIOTests/EventLoopTest.swift]: Can we add support for joining `Thread` instead? I'd vastly prefer having a deterministic test than one that has an infinite loop to avoid timing issues.

Comment (normanmaurer) [reply Tests/NIOTests/EventLoopTest.swift]: I can do this sure

Comment (weissi) [reply Tests/NIOTests/EventLoopTest.swift]: sounds like the right thing to do...

**Comment by helje5** [comment]
Does someone know why `dispatch_get_current_queue()` got dropped? There must have been a very good reason for this, which might also apply to NIO event loops? Just wondering.
(I know that OperationQueues still have it).

**Comment by weissi** [comment]
@helje5 various reasons:

- you can be on multiple queues (see target queues)
- `dispatch_get_current_queue()` might crash your program if used under ARC (as the returned queue might have reference count 0 already so ARC would resurrect it)
- generally `DispatchQueues` and our `EventLoop` look similar but aren't really:
  - An `EventLoop` is always tied to a thread, a `DispatchQueue` is not
  - You can only ever be on one `EventLoop` but you can be 'on' multiple `DispatchQueue`s

File: Sources/NIO/Thread.swift [hunk 6]
```diff
@@ -117,6 +117,11 @@ final class Thread {
         return pthread_equal(pthread, pthread_self()) != 0
     }
 
+    func join() {
+        // We ignore the return value as the pthread may already be dead.
+        pthread_join(pthread, nil)
```
Comment (Lukasa) [inline Sources/NIO/Thread.swift]: Should we add a precondition here that confirms that we aren't joining on ourselves?

Comment (toffaletti) [reply Sources/NIO/Thread.swift]: If you are adding support for joining threads, shouldn't you also stop detaching in spawnAndRun?

Comment (normanmaurer) [reply Sources/NIO/Thread.swift]: Let me make it an assert

Comment (normanmaurer) [reply Sources/NIO/Thread.swift]: @toffaletti good Point....  @jw @Lukasa agree ?

Comment (weissi) [reply Sources/NIO/Thread.swift]: I'm certainly happy with this change & `assert`. Should be documented then (and also `preconditioned` in `Thread.deinit` that the thread has been joined

Comment (weissi) [reply Sources/NIO/Thread.swift]: actually, that has one problem (which is certainly API breaking): previously we didn't actually require you to shut down the event loops (as we had detached threads and we never joined). With that change we would require users to shut the event loops down properly in order for us to join the threads. And it's required to join non-detached threads or else Thread sanitizer will barf.

Comment (normanmaurer) [reply Sources/NIO/Thread.swift]: you mean the check in deinit would be a breakage currect @weissi ? Inthinkwe can find everything except the deinit change

Comment (normanmaurer) [reply Sources/NIO/Thread.swift]: Beside this we could also just revert the join but and use the while loop in the test as before. At the end it was mainly to make the test better

Comment (weissi) [reply Sources/NIO/Thread.swift]: @normanmaurer yes. If we decide to not detach the threads anymore, we have two options (both of which break):

1) we do put a `precondition` in `deinit` that a `Thread` has been `join`ed before being released. That'd break as currently you're not required to do that.
2) we do _not_ put a `precondition` in `deinit`: That has the same effect when run under TSan as TSan barks if you leaks a thread that is running (unless it's detached).

In other words, if we don't detach our threads we have a problem as we currently don't require users to shut down the event loops (which is a mistake).

Comment (weissi) [reply Sources/NIO/Thread.swift]: oh, and as @toffaletti points out, we can't `pthread_join` a thread that is detached. So I think we're stuck 😤

Comment (weissi) [reply Sources/NIO/Thread.swift]: @normanmaurer yes, I think we have to back off with the joining :|. We should however make this change in 2.0.0.

Comment (normanmaurer) [reply Sources/NIO/Thread.swift]: Ok let me use the while loop again and add a comment that we should use join in the next major

**Comment by normanmaurer** [comment]
@weissi @toffaletti @Lukasa PTAL again... as discussed I did remove the join bits https://github.com/apple/swift-nio/pull/120#discussion_r174239321

File: Sources/NIO/EventLoop.swift [hunk 7]
```diff
@@ -729,6 +742,9 @@ final public class MultiThreadedEventLoopGroup: EventLoopGroup {
 
         g.notify(queue: q) {
             let failure = self.eventLoops.map { try? $0.close0() }.filter { $0 == nil }.count > 0
+
+            // TODO: In the next major release we should join in the Thread used by the EventLoop before invoking the callback to ensure
+            //       it is really gone.
```
Comment (toffaletti) [inline Sources/NIO/EventLoop.swift]: No need to change  this PR, but `ThreadSpecificVariable` is also going to need to change to properly call pthread_key_delete inside the destructor function passed to pthread_key_create. I ran into this working on #133 but couldn't wrap my head around how to fix it.

Comment (normanmaurer) [reply Sources/NIO/EventLoop.swift]: @weissi can you take a look ?

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: @toffaletti / @normanmaurer using `pthread_key_delete` is unfortunately not straightforward at all I think. I'll create an issue.

**Comment by Lukasa** [comment]
Rebase and then merge away.

**Comment by normanmaurer** [comment]
@Lukasa Roger that 👍
